Tempo: Enhance gRPC client with full OpenTelemetry tracing and Prometheus metrics

QuentinBisson opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

This commit improves the Tempo datasource's gRPC client to achieve feature parity with the HTTP client by implementing production-ready observability:

Changes:
- Add full OpenTelemetry tracing with proper span attributes and error handling
- Implement comprehensive Prometheus metrics (requests, duration, in-flight)
- Centralize user agent handling via interceptor instead of manual setting
- Add proper metric initialization with thread-safe lazy loading
- Remove manual user agent code from streaming functions

Why this is needed:
- The gRPC client previously only had basic logging-based interceptors
- HTTP client had full observability stack while gRPC client lacked monitoring
- Production deployments require comprehensive metrics and tracing for:
  * Request success/failure rates and error tracking
  * Performance monitoring via duration histograms
  * Concurrent request tracking for capacity planning
  * Distributed tracing for request correlation across services

The enhanced gRPC client now provides the same level of production observability as the HTTP client while maintaining streaming performance.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

QuentinBisson committed:
Tempo: Enhance gRPC client with full OpenTelemetry tracing and Prometheus metrics

This commit improves the Tempo datasource's gRPC client to achieve feature
parity with the HTTP client by implementing production-ready observability:

Changes:
- Add full OpenTelemetry tracing with proper span attributes and error handling
- Implement comprehensive Prometheus metrics (requests, duration, in-flight)
- Centralize user agent handling via interceptor instead of manual setting
- Add proper metric initialization with thread-safe lazy loading
- Remove manual user agent code from streaming functions

Why this is needed:
- The gRPC client previously only had basic logging-based interceptors
- HTTP client had full observability stack while gRPC client lacked monitoring
- Production deployments require comprehensive metrics and tracing for:
  * Request success/failure rates and error tracking
  * Performance monitoring via duration histograms
  * Concurrent request tracking for capacity planning
  * Distributed tracing for request correlation across services

The enhanced gRPC client now provides the same level of production
observability as the HTTP client while maintaining streaming performance.

QuentinBisson committed:
Merge branch 'main' into improve-tempo-datasource

QuentinBisson commented:
It's not urgent, sorry, I was going through my opened PRs :D

QuentinBisson committed:
Clean up CHANGELOG by removing unreleased section

Removed unreleased section and feature details from CHANGELOG.

QuentinBisson committed:
Merge branch 'main' into improve-tempo-datasource

QuentinBisson committed:
Merge branch 'main' into improve-tempo-datasource

QuentinBisson commented on pkg/tsdb/tempo/grpc.go:
```diff
@@ -101,10 +161,16 @@ func getDialOpts(ctx context.Context, settings backend.DataSourceInstanceSetting
 			maxCallRecvMsgSizeBytes = parsed
 		}
-	}
+  }
```
```suggestion
    }
```

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson commented on pkg/tsdb/tempo/grpc.go:
```diff
@@ -101,10 +161,16 @@ func getDialOpts(ctx context.Context, settings backend.DataSourceInstanceSetting
 			maxCallRecvMsgSizeBytes = parsed
 		}
-	}
+    }
```
```suggestion
	}
```

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson commented on pkg/tsdb/tempo/grpc.go:
```diff
@@ -8,18 +8,71 @@ import (
 	"google.golang.org/grpc/metadata"
 
 	"github.com/grafana/grafana-plugin-sdk-go/backend"
 	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
```
```suggestion
	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
	"github.com/grafana/grafana-plugin-sdk-go/backend/tracing"
```

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson commented on pkg/tsdb/tempo/grpc.go:
```diff
@@ -8,18 +8,72 @@ import (
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/grafana/grafana/pkg/infra/tracing"
```
```suggestion
```

QuentinBisson committed:
Apply suggestion from @QuentinBisson

QuentinBisson commented on pkg/tsdb/tempo/grpc.go:
```diff
@@ -170,6 +235,85 @@ func CustomHeadersStreamInterceptor(httpOpts httpclient.Options) grpc.StreamClie
+func TracingStreamInterceptor() grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		// Start an OpenTelemetry span for the gRPC call
+		ctx, span := tracing.Start(ctx, "tempo.grpc.stream", 
```
```suggestion
		ctx, span := tracing.DefaultTracer().Start(ctx, "tempo.grpc.stream", 
```

QuentinBisson committed:
Apply suggestion from @QuentinBisson

itsmylife committed:
Merge branch 'main' into improve-tempo-datasource

itsmylife reviewed (changes requested):
I wanted to fix the conflicts and review it but it went side ways. Sorry about this. 
Could you please fix the conflicts and address this issue?

```
cannot use attribute.Bool("stream.client", true) (value of struct type "go.opentelemetry.io/otel/attribute".KeyValue) as "go.opentelemetry.io/otel/trace".SpanStartOption value in argument to tracing.DefaultTracer().Start: "go.opentelemetry.io/otel/attribute".KeyValue does not implement "go.opentelemetry.io/otel/trace".SpanStartOption (missing method applySpanStart)
```

QuentinBisson committed:
Merge branch 'main' into improve-tempo-datasource

QuentinBisson committed:
fix conflict

QuentinBisson committed:
fix-linting

jcolladokuri reviewed (commented):
Thanks for the contribution! LGTM, just a few minor nits to point out :)

itsmylife reviewed (changes requested):
@QuentinBisson two small things for you to handle and then we can just merge this. Thank you for your time!

QuentinBisson committed:
Tempo: address review feedback on gRPC interceptors

- TracingStreamInterceptor: return backend.DownstreamErrorf instead of
  logging and passing through the raw error
- MetricsStreamInterceptor: use real gRPC status code names (OK,
  UNAVAILABLE, …) as metric labels instead of "success"/"error"
- Add grpcStatusLabel helper using google.golang.org/grpc/status

itsmylife reviewed (approved):
🎉
